Replace libdparse in UnusedResultChecker

Vladiwostok opened the pull request:
https://github.com/Dlang-UPB/D-scanner/pull/74 cont

Vladiwostok committed:
Replace libdparse with DMD in UnusedResultChecker

RazvanN7 commented on src/dscanner/analysis/unused_result.d:
```diff
@@ -24,174 +16,186 @@ import std.range : retro;
+            fun(); // [warn]: %s
         }
-    }c, sac);
+    }c.format(MSG), sac, 1);
```
I would also leave the `void func` test just to make sure that we do not have any false positives.

RazvanN7 reviewed (commented):
It seems that you are modifying the tests extensively. Could you please provide an explanation on why that is the case?

RazvanN7 commented on src/dscanner/analysis/unused_result.d:
```diff
@@ -24,174 +16,186 @@ import std.range : retro;
+		}
+
+		super.visit(ifStatement);
+	}
```
I don't understand why you need separate visitation methods for ifs and compound statements. This is more elegantly solved by simply transforming the `hasUnusedResult` function into a generic statement visitation function that simply calls `super.visit()` on the statement parameter at the end of the function.

RazvanN7 replied:
Ah nvm that. I understand now.

RazvanN7 reviewed (changes requested):
The code is too complex for what we need to do. Refactor by using a single visitation method for statements that calls `super.visit`.

RazvanN7 reviewed (approved):
The implementation is OK, now that I better understand what you are doing. But please mind the tests.

Vladiwostok commented:
[DO NOT MERGE] I think one unit test has been slightly modified after the rebase from Upstream D-scanner. I'll do an extra verification and make sure everything is OK - in the meantime please do not merge this PR

Vladiwostok commented:
> The implementation is OK, now that I better understand what you are doing. But please mind the tests.

> [DO NOT MERGE] I think one unit test has been slightly modified after the rebase from Upstream D-scanner. I'll do an extra verification and make sure everything is OK - in the meantime please do not merge this PR

The unit tests have changed in the upstream in the last year - that's why the big diff on unit test. Will update & fix

Vladiwostok committed:
Run semantic analysis in unit test

Vladiwostok committed:
Add more unit tests

Vladiwostok commented:
> > The implementation is OK, now that I better understand what you are doing. But please mind the tests.
> 
> > [DO NOT MERGE] I think one unit test has been slightly modified after the rebase from Upstream D-scanner. I'll do an extra verification and make sure everything is OK - in the meantime please do not merge this PR
> 
> The unit tests have changed in the upstream in the last year - that's why the big diff on unit test. Will update & fix

Fixed

Vladiwostok committed:
Remove debug prints

Vladiwostok committed:
Fix no return test

Vladiwostok commented on src/dscanner/analysis/unused_result.d:
```diff
@@ -24,123 +16,136 @@ import std.range : retro;
         {
             fun();
         }
-    }c, sac);
```
@RazvanN7  There's a divergence between running unit tests and running d-scanner.
assertAnalyzerWarningsDMD requires full semantic analysis for this check's unittest to pass.
I've found this in run.d: https://github.com/Dlang-UPB/D-scanner/blob/replace_libdparse/src/dscanner/analysis/run.d#L421 .
Seems that full semantic is commented out; is there any reason?
Why is there a possibility to run full semantic in unit tests? Shouldn't that also be removed?

RazvanN7 commented:
This issue is in the 'Ready for review' queue, however, it seems there are unaddressed issues. Can you clarify whether this is ready for review or not?

Vladiwostok commented:
> This issue is in the 'Ready for review' queue, however, it seems there are unaddressed issues. Can you clarify whether this is ready for review or not?

Ready for review

RazvanN7 commented:
I find it kind of weird that this is passing all tests (including testing phobos) since you essentially need to perform semantic analysis for this check to work properly and it seems that you are only doing semantic when testing unittests. Is that right?

RazvanN7 commented:
Mind the conflict.

RazvanN7 commented:
ping

Vladiwostok commented:
> ping

On hold for now.
https://github.com/Dlang-UPB/D-scanner/issues/102 has to be implemented first

Vladiwostok committed:
Remove semantic analysis

Vladiwostok commented on src/dscanner/analysis/unused_result.d:
```diff
@@ -24,105 +16,212 @@ import std.range : retro;
+	}
 
-public:
+	private template VisitAggregate(NodeType)
```
This is necessary in order to keep the exact same behavior as the previous implementation with libdparse
